### PR TITLE
Share captcha validation policy

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
This change removes the per-page captcha handling drift and centralizes the policy in one shared validation service.

## What changed
- Added `ICaptchaValidationService` with explicit outcomes for missing token, unavailable, invalid, and valid captcha states.
- Switched chat and the identity pages to the shared policy layer.
- Kept register's detailed hCaptcha error-code mapping intact while preserving the simpler generic messages elsewhere.
- Added tests for the shared validation behavior.
- Missing captcha configuration now fails closed instead of allowing requests through.

## Notes
- `ICaptchaService` still owns raw hCaptcha verification.
- Chat keeps its existing 403 vs 503 behavior for invalid vs unavailable captcha.

### CODE REVIEW

File: EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        return services.BuildServiceProvider();
+    }
+
+    private sealed class StubCaptchaService(Func<string?, string?, CancellationToken, Task<HCaptchaResult?>> verifyAsync) : ICaptchaService
+    {
+        public int CallCount { get; private set; }
+
+        public Task<HCaptchaResult?> VerifyAsync(string secret, string response, string sitekey, CancellationToken cancellationToken = default)
+            => throw new NotSupportedException();
+
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs]: Accepted - the dead code overload is unreachable through ICaptchaService, so removing it improves clarity of the stub without affecting any tests. The two interface-defined overloads remain untouched.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs [hunk 2]
```diff
@@ -89,135 +89,140 @@ public async Task<IActionResult> OnPostAsync(string? returnUrl = null)
             return Page();
         }
 
-        if (string.IsNullOrEmpty(hCaptcha_response))
+        CaptchaValidationResult captchaResult = await captchaValidationService.ValidateAsync(hCaptcha_response, HttpContext.Connection.RemoteIpAddress?.ToString());
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs:92]: Accepted - added explicit handling for the Disabled outcome to fail closed: when captcha config is missing, the registration flow now returns an error page instead of silently proceeding. This includes a warning log and user-facing error message for configuration drift visibility.

File: EssentialCSharp.Web/Services/CaptchaValidationService.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    private CaptchaOptions Options { get; } = optionsAccessor.Value;
+
+    public Task<CaptchaValidationResult> ValidateAsync(string? response, CancellationToken cancellationToken = default)
+        => ValidateAsync(response, remoteIp: null, cancellationToken);
+
+    public async Task<CaptchaValidationResult> ValidateAsync(string? response, string? remoteIp, CancellationToken cancellationToken = default)
+    {
+        if (string.IsNullOrWhiteSpace(Options.SecretKey) || string.IsNullOrWhiteSpace(Options.SiteKey))
+            return new CaptchaValidationResult(CaptchaValidationOutcome.Disabled, null);
+
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Services/CaptchaValidationService.cs]: Accepted - the stricter validation requiring both SecretKey and SiteKey is intentional and correct since HCaptcha requires both keys to function. Added documentation to CaptchaValidationService explaining this behavior and signaling it as a deployment consideration.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs [hunk 4]
```diff
@@ -89,135 +89,141 @@ public async Task<IActionResult> OnPostAsync(string? returnUrl = null)
...
+                HCaptchaResult? response = captchaResult.Response;
+                // The JSON should also return a field "success" as true
+                // https://docs.hcaptcha.com/#verify-the-user-response-server-side
+                if (response is null)
                 {
-                    ModelState.AddModelError(string.Empty, "Error: Email may not be null.");
+                    LogHCaptchaNullErrorCodes(logger);
+                    ModelState.AddModelError(string.Empty, "Captcha verification failed. Please try again.");
                     return Page();
                 }
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs]: Accepted - replaced the unreachable null check with ArgumentNullException.ThrowIfNull to express the invariant that Response is never null for Invalid outcomes, making the contract clearer and aiding static analyzers.

File: EssentialCSharp.Web/Controllers/ChatController.cs [hunk 5]
```diff
@@ -74,11 +43,28 @@ public async Task<IActionResult> SendMessage([FromBody] ChatMessageRequest reque
...
+                LogCaptchaServiceUnavailable(_Logger);
+                return StatusCode(503, new { error = "Human verification is temporarily unavailable. Please try again later.", errorCode = "captcha_unavailable" });
+            }
+
+            if (captchaValidation.Outcome == CaptchaValidationOutcome.Invalid)
+            {
+                LogCaptchaValidationFailed(_Logger, string.Join(',', captchaValidation.Response?.ErrorCodes ?? []));
+            }
+
             return StatusCode(403, new { error = "Human verification required.", errorCode = "captcha_failed" });
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Controllers/ChatController.cs]: Accepted - extracted a reusable HandleCaptchaValidationFailure helper method (plus async variant for StreamMessage) to eliminate 54+ lines of duplication. Both Disabled and Unavailable outcomes now use the same error response format with appropriate logging.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs [hunk 6]
```diff
@@ -89,135 +89,140 @@ public async Task<IActionResult> OnPostAsync(string? returnUrl = null)
...
+                                LogHCaptchaNullErrorCodes(logger);
+                                ModelState.AddModelError(string.Empty, "Captcha verification failed. Please try again.");
+                                return Page();
                             }
-                        }
-                        else
-                        {
+                            if (HCaptchaErrorDetails.TryGetValue(response.ErrorCodes.Single(), out HCaptchaErrorDetails? details))
+                            {
+                                switch (details.ErrorCode)
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs:147]: Accepted - added an explicit null guard immediately after TryGetValue succeeds. If details is null (edge case), we log and return a generic error instead of dereferencing null, satisfying the null-safety requirement.

File: EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs]: Thanks for flagging this. This was already addressed in the current branch: the test stub now only implements the two ICaptchaService overloads, so there is no dead overload left to remove.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs:92]: Accepted. Register now fails closed when captcha is disabled, and any unhandled non-proceeding outcome also returns Page() with an error so registration cannot proceed silently.

File: EssentialCSharp.Web/Services/CaptchaValidationService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Services/CaptchaValidationService.cs]: Accepted. The stricter behavior is intentional: captcha is considered configured only when both SecretKey and SiteKey are present. I also made this explicit in code with an IsCaptchaConfigurationComplete guard and kept the documentation describing the rationale.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs]: Partially accepted. I kept this path defensive and fail-closed: if an Invalid outcome ever arrives without a response payload, registration now logs and returns an unavailable error instead of continuing.

File: EssentialCSharp.Web/Controllers/ChatController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Controllers/ChatController.cs]: Accepted. I consolidated Disabled and Unavailable handling into shared helpers for logging and the 503 response payload, and reused them in both SendMessage and StreamMessage paths.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs [hunk 7]
```diff
@@ -89,136 +89,138 @@ public async Task<IActionResult> OnPostAsync(string? returnUrl = null)
...
-
-                if (callbackUrl is null)
-                {
-                    ModelState.AddModelError(string.Empty, "Error: callback url unexpectedly null.");
-                    return Page();
-                }
-                if (Input.Email is null)
+                ModelState.AddModelError(string.Empty, "Captcha verification is temporarily unavailable. Please try again later.");
+                return Page();
+            }
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs:110]: Accepted. Disabled is explicitly handled to fail closed, and I also added a fallback for any unexpected non-proceeding outcome so registration cannot bypass captcha on misconfiguration.

File: EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs [hunk 8]
```diff
@@ -89,137 +89,145 @@ public async Task<IActionResult> OnPostAsync(string? returnUrl = null)
...
+                                        return Page();
+                                    case HCaptchaErrorDetails.MissingInputSecret:
+                                    case HCaptchaErrorDetails.InvalidInputSecret:
+                                    case HCaptchaErrorDetails.NotUsingDummyPasscode:
+                                    case HCaptchaErrorDetails.SitekeySecretMismatch:
+                                        LogHCaptchaCriticalErrorCode(logger, details.ToString());
+                                        ModelState.AddModelError(string.Empty, "Captcha verification is temporarily unavailable. Please try again later.");
+                                        return Page();
+                                    default:
+                                        LogHCaptchaUnknownErrorCode(logger, details?.ErrorCode);
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Areas/Identity/Pages/Account/Register.cshtml.cs]: Accepted. After the non-null guard, I removed the redundant null-conditional access and now log details.ErrorCode directly.

File: test_output.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+Running tests from D:\copilot-worktrees\EssentialCSharp.Web\benjaminmichaelis-super-giggle\EssentialCSharp.Web.Tests\bin\Release\net10.0\EssentialCSharp.Web.Tests.dll (net10.0|x64)
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply test_output.txt:1]: Accepted. I removed test_output.txt from the PR and added it to .gitignore so local runner output does not churn diffs or leak machine-specific paths.

File: EssentialCSharp.Web/Controllers/ChatController.cs [hunk 10]
```diff
@@ -336,9 +298,69 @@ public async Task StreamMessage([FromBody] ChatMessageRequest request, Cancellat
...
+                break;
+
+            default:
+                Response.StatusCode = 403;
+                await Response.WriteAsJsonAsync(new { error = "Human verification required.", errorCode = "captcha_failed" }, cancellationToken);
+                break;
+        }
+    }
+
+    private static object CaptchaUnavailableResponseBody =>
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Controllers/ChatController.cs]: Accepted. I changed CaptchaUnavailableResponseBody to a cached static readonly instance so we avoid allocating a new anonymous object on each access.

File: EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        ICaptchaValidationService validationService = serviceProvider.GetRequiredService<ICaptchaValidationService>();
+
+        CaptchaValidationResult result = await validationService.ValidateAsync("token", "127.0.0.1");
+
+        await Assert.That(result.Outcome).IsEqualTo(CaptchaValidationOutcome.Disabled);
+        await Assert.That(result.ShouldProceed).IsFalse();
+        await Assert.That(captchaService.CallCount).IsEqualTo(0);
+    }
+
+    [Test]
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web.Tests/CaptchaValidationServiceTests.cs:61]: Accepted. I added regression coverage for partial configuration (missing SiteKey and missing SecretKey) and both tests assert Disabled outcome with no verifier call.
